### ISSUE TITLE
[B] Apply mods and keep shape keys gives wrong result

### ISSUE THREAD

**Issue by schroef** [JanOtt/ShapeKey-Helpers#1]
v1.0.0
bl 2.79c
osx 10.11.6

I just found this addon and mainly wanted to test it for a quite a basic object. Its not anything with an armature thought.

I have this tube kind of shape with some mods on it (mirror, subs, solidify) i got a base shape key and 2 sets. When i do the action i see a new shape with no mods, no shape keys. 

I think the issue is because i already have a copy of that mesh in my scene. I tend to keep the filenaming of .001 because it works and sometimes helps.

Mesh with mods and shape keys
<img width="590" alt="org-mesh" src="https://user-images.githubusercontent.com/6923008/47518952-589adb00-d85a-11e8-9099-0bed930a7199.png">

Result mesh in viewport
<img width="603" alt="result-mesh" src="https://user-images.githubusercontent.com/6923008/47518951-589adb00-d85a-11e8-96a1-b51918db2be3.png">

edit
My thoughts were this was causing issues with this operator. But after changing is till dont see shape keys back no the new shape. When i check the list of shapes i do see the created meshes. I dont get errors in the console. I know the easy fix is select the shapes than the new object and choose add ass shape keys.

Just wanted to post this bug
Im not sure what mods could cause an issue, i guess the solidify is causing an issue here. The Blenderartist link is at a deadend in the main page.

Adjust meshlist, doesnt help 
<img width="259" alt="mesh list-new" src="https://user-images.githubusercontent.com/6923008/47518949-58024480-d85a-11e8-8f48-2dea3754fbef.png">

Adjust meshlist, i though already excisting copy was causing issues here
<img width="257" alt="mesh list-old" src="https://user-images.githubusercontent.com/6923008/47518950-58024480-d85a-11e8-80c1-0dab0855554d.png">

**Comment by JanOtt** [comment]
Hi schroef,

I'm not really actively working on this addon anymore, but thanks for reaching out!

Sad to see that the Blenderartists topic is gone (probably due to switching to a new forum system), there was a lot of helpful advice in there. 

Anyway, I think I know what causes your problem and just modified and updated the 'shapeKeyHelper_menu.py' file. Try downloading the new file, replace your old .py file and check if the shapekeys are added correctly this time.

**Comment by schroef** [comment]
I think also because i did used a armature and also used solidify mod. Perhaps some features of taking these off or adding them later can be added?

When i added a armature (nothing set) mod it did work. Checking the code, it seems this function first check if there is an armature. I wanted to use it on a simple model, not a character.

**Comment by JanOtt** [comment]
Yes, using a 'Solidify' modifier would have prevented the addon from working properly before. That's fixed now, though. 

Whether you have an armature attached to your object or not should make no difference. The script only checks for an armature to disable it while the shapekeys are being handled. If there's no armature on your object at all, that step is simply skipped, but the rest of the code is still executed.

If the addon still doesn't work, you could try sending me a .blend file of your project and I'll take a closer look.

**Comment by schroef** [comment]
Works like a charm now! Perhaps it would be nice if it sets the shape keys back to their original value. Now it leaves the original object shapes keys at 0 and also on the "applied" object.

Its not that hard to copy the original settings and reapply them.

EDIT
after checking your description in the "shape_key_preserver", it states "Creates a new object with all suitable modifiers applied (Subdivision Surface, Mirror, Bevel and Edgesplit) and all shape keys preserved". Could it be an option that it also adds back the mods which are not in the list of applied mods?

I also get this "error' in the console about the duplicated meshes
"DM_to_mesh: YEEK! this should be recoded! Shape key loss!: ID 'METhudnerstick_base-tube.002'
"

Im not sure why the shape value is first cleared and right after it it set back to 1? Thought it seems its not set to 1 because after the operator is done the values are all 0

See line #147 and #149
```
bpy.ops.object.shape_key_clear()
            
shapekey.value = 1.0
```

**Comment by schroef** [comment]
Here's quick dirty edit which also applies the same value to the shape keys and also puts these on the new mesh and also set back the old values on the original mesh.

```
        listOfShapeInstances = []
        shapekeyOrg = []
        #_______________________________________________________________

        #Deactivate any armature modifiers
        for mod in originalObject.modifiers:
            if mod.type == 'ARMATURE':
                originalObject.modifiers[mod.name].show_viewport = False

        index = 0
        for shapekey in originalObject.data.shape_keys.key_blocks:
            if(index == 0):
                index = index + 1
                continue
            shapekeyOrg.append(shapekey.value)

        index = 0
        for shapekey in originalObject.data.shape_keys.key_blocks:

            if(index == 0):
                index = index + 1
                continue

            bpy.ops.object.select_all(action='DESELECT')
            originalObject.select = True
            bpy.context.scene.objects.active = originalObject

            bpy.ops.object.shape_key_clear()

            shapekey.value = 1.0

            #save name
            #____________________________
            shapekeyname = shapekey.name

            #create new object from shapekey and add it to list
            #____________________________
            bpy.ops.object.duplicate(linked=False, mode='TRANSLATION')
            bpy.ops.object.convert(target='MESH')
            listOfShapeInstances.append(bpy.context.active_object)

            #rename new object
            #____________________________
            bpy.context.object.name = shapekeyname



            bpy.ops.object.select_all(action='DESELECT')
            originalObject.select = True
            bpy.context.scene.objects.active = originalObject

        #_____________________________________________________________
        #Prepare final empty container model for all those shape keys:
        #_____________________________________________________________

        bpy.context.scene.objects.active = originalObject
        bpy.ops.object.shape_key_clear()

        bpy.ops.object.duplicate(linked=False, mode='TRANSLATION')
        newObject = bpy.context.active_object

        bpy.ops.object.shape_key_clear()
        bpy.ops.object.shape_key_remove(all=True)

        newObject.name = oldName + "_Applied"

        for mod in newObject.modifiers:
            bpy.ops.object.modifier_apply(apply_as='DATA', modifier=mod.name)

        for object in listOfShapeInstances:

            bpy.ops.object.select_all(action='DESELECT')
            newObject.select = True
            object.select = True
            bpy.context.scene.objects.active = newObject
            bpy.ops.object.join_shapes()

        #put back original value to new shape
        index = 0
        for shapekey in newObject.data.shape_keys.key_blocks:
            if(index == 0):
                index = index + 1
                continue
            shapekey.value = shapekeyOrg[index-1]
            index = index + 1

        #put back original value to old shape
        index = 0
        for shapekey in originalObject.data.shape_keys.key_blocks:
            if(index == 0):
                index = index + 1
                continue
            shapekey.value = shapekeyOrg[index-1]
            index = index + 1
```

**Comment by JanOtt** [comment]
Sorry for taking so long to answer, unfortunately I was very busy recently.

Anyway, thanks for your 'quick edit' (which works perfectly). I added your code to the addon and the updated version is now online.

---

> after checking your description in the "shape_key_preserver", it states "Creates a new object with all suitable modifiers applied (Subdivision Surface, Mirror, Bevel and Edgesplit) and all shape keys preserved". Could it be an option that it also adds back the mods which are not in the list of applied mods?

I'm sorry for that misleading description in the previous version of the addon. The operator doesn't really differentiate between 'suitable' and 'non-suitable' modifiers. 

All it really does is creating a separate object for each shapekey, applies all active modifiers on each of these objects and then tries to combine all objects back into one.

This procedure only works if all of these generated "shape key objects" have the same number of vertices (or else Blender can't combine them). This means that modifiers like _Subdivision Surface_ will always work, while other modifiers will work most likely (_Mirror_) and others may only work under specific circumstances (_Edge Split_). Modifiers like _Remesh_ will most definitely never work.

You can check if the operator is going to work correctly by checking the _verts_ count (number of vertices) of your scene in the upper right corner of the default UI preset while changing the value of your shapekeys on your objects. If the _verts_ count stays the same, no problems. If it **does**, the operator isn't going to work.

> I also get this "error' in the console about the duplicated meshes
> "DM_to_mesh: YEEK! this should be recoded! Shape key loss!: ID 'METhudnerstick_base-tube.002'"

That's a harmless (yet annoying) warning you don't really have to worry about. Blender throws it every time you use the 'Convert to Mesh/Text' operator on an object with shape keys. I use this operator in the addon to apply all shapekeys on the multiple generated "shape key objects" I mentioned earlier.

In other words, the addon is working exactly like it's supposed to.

> Im not sure why the shape value is first cleared and right after it it set back to 1? Thought it seems its not set to 1 because after the operator is done the values are all 0

This step is simply done to make sure that only **one** shape key is deforming the mesh, before copying the whole object to create one of the "shape key objects" mentioned above. 

---

I hope my explanations are at least somewhat understandable. Explaining all of this in text is very difficult.

**Comment by schroef** [comment]
Yes, more clear now. Im glad i could help a bit

**Comment by schroef** [comment]
PS shouldn't you update the version number? I also noticed that you use releases, this one states v1.1?

Here's good article for keeping change log, i started doing this and makes it very clear what has been changed etc etc
https://keepachangelog.com/en/1.0.0/
